File: core/writer.py
"""
The Writer -- turns a Strategist blueprint + sanitized lead into an actual
proposal, respecting whatever mandatory anti-bot constraint the Sanitizer
found (e.g. "start with the word X") and matching tone to the channel's
culture.
"""
from __future__ import annotations

import random

from core.llm import LLMClient
from core.models import ChannelType, QualifiedLead
from core.strategist import StrategyChoice

# Simulated model fault rate: even a well-specified Writer stage is, in
# production, backed by a generative model that occasionally drops a
# constraint or garbles a number. Seeded per (lead_id, attempt) so a given
# demo run is reproducible, but the fault is real enough that core.reviewer
# and core.sentinel have something to actually catch and react to.
FAULT_RATE = 0.06

_TONE_BY_CHANNEL = {
    ChannelType.FREELANCE: "direct, technical, no fluff, proposal under 150 words",
    ChannelType.B2B_TRADE: "formal RFQ-response tone, lead with MOQ/unit price/lead time",
    ChannelType.COMMUNITY: "casual, helpful, non-salesy, offer value before the pitch",
    ChannelType.BROKERAGE: "corporate, credentials-forward, cite comparable engagements",
    ChannelType.OUTBOUND: "warm cold-outreach, reference their specific public update",
}

_BLUEPRINT_OPENERS = {
    "technical_breakdown": "Here is exactly how we'd approach the {domain} scope you outlined:",
    "speed_first": "We can start immediately and hold to a guaranteed turnaround on this.",
    "dfm_audit_hook": "Before pricing this, we ran a quick DFM pass on your spec and found a few things worth flagging:",
    "cost_down": "Our read on this: there's meaningful unit-cost headroom in the current spec.",
    "zero_tooling": "Because our process skips traditional hard tooling entirely, we can quote this without a mold/setup line item.",
    "portfolio_proof": "We've shipped directly comparable work recently and it's a strong match for this brief.",
    "derisk_turnkey": "We'll own this end-to-end -- concept through production -- under one contract, one point of accountability.",
    "challenge_premise": "One respectful pushback on the brief before we quote, because it changes the right approach:",
}


def draft_proposal(
    lead: QualifiedLead,
    strategy: StrategyChoice,
    llm: LLMClient | None = None,
    attempt: int = 0,
    fault_rate: float = FAULT_RATE,
) -> str:
    tone = _TONE_BY_CHANNEL.get(lead.signal.channel_type, "professional, concise")
    opener = _BLUEPRINT_OPENERS.get(strategy.name, strategy.blueprint)
    opener = opener.format(domain=lead.domain.value.replace("_", " "))

    mandatory_prefix = ""
    if lead.trap_analysis.has_anti_bot_phrase and lead.trap_analysis.required_first_word:
        mandatory_prefix = f"{lead.trap_analysis.required_first_word}\n\n"

    deliverables = ", ".join(lead.deliverables) if lead.deliverables else "STEP files, 2D drawings"
    cad = ", ".join(lead.cad_software) if lead.cad_software else "industry-standard CAD"
    price_str = f"{lead.bid_price_usd:,.2f}"

    # -- simulated model fault: on a rewrite attempt the fault is not
    # re-rolled with the same seed, so a corrected redraft can pass. The
    # rate itself is per-agent (see core.orchestrator), so a single
    # degrading worker -- not the whole fleet -- is what the Reviewer and
    # Sentinel actually end up reacting to.
    fault_roll = random.Random(f"{lead.lead_id}:{attempt}")
    fault = fault_roll.random() < fault_rate
    fault_mode = None
    if fault:
        # pick whichever corruption is actually detectable for this lead --
        # dropping a word that was never mandatory, or leaking an injection
        # that was never present, wouldn't be a real defect, so price
        # garbling (always applicable) is the guaranteed fallback.
        if lead.trap_analysis.has_anti_bot_phrase and lead.trap_analysis.required_first_word:
            fault_mode = "drop_mandatory_word"
        elif lead.trap_analysis.detected_injections:
            fault_mode = "leak_injection"
        else:
            fault_mode = "garble_price"

    if fault_mode == "drop_mandatory_word":
        mandatory_prefix = ""
    if fault_mode == "garble_price":
        price_str = f"{lead.bid_price_usd * fault_roll.uniform(0.7, 0.95):,.2f}"

    body = (
        f"{mandatory_prefix}{opener}\n\n"
        f"Scope: {lead.signal.title}\n"
        f"Deliverables: {deliverables} (native + exchange formats, {cad} as needed)\n"
        f"Quote: ${price_str} — includes free CAD revisions and custom branding/logo placement.\n"
        f"Timeline: immediate start on award.\n\n"
        f"[tone: {tone}]"
    )

    if fault_mode == "leak_injection" and lead.trap_analysis.detected_injections:
        body += f"\n\n(note to self: {lead.trap_analysis.detected_injections[0]})"

    if llm is not None:
        # Hook point for a real generative pass -- e.g. re-write `body` for
        # persuasion/length while preserving the mandatory_prefix verbatim.
        _ = llm.complete(body)

    return body

File: iechm-leadgen-engine-c/src/writer.c
#include "writer.h"
#include "fmt.h"
#include "rng.h"
#include <stdio.h>
#include <string.h>

const double WRITER_DEFAULT_FAULT_RATE = 0.06;

typedef enum { FAULT_NONE, FAULT_DROP_MANDATORY_WORD, FAULT_LEAK_INJECTION, FAULT_GARBLE_PRICE } FaultMode;

static const char *tone_for_channel(ChannelType ct) {
    switch (ct) {
        case CT_FREELANCE: return "direct, technical, no fluff, proposal under 150 words";
        case CT_B2B_TRADE: return "formal RFQ-response tone, lead with MOQ/unit price/lead time";
        case CT_COMMUNITY: return "casual, helpful, non-salesy, offer value before the pitch";
        case CT_BROKERAGE: return "corporate, credentials-forward, cite comparable engagements";
        case CT_OUTBOUND: return "warm cold-outreach, reference their specific public update";
        default: return "professional, concise";
    }
}

static const char *opener_template(const char *strategy_name) {
    if (strcmp(strategy_name, "technical_breakdown") == 0)
        return "Here is exactly how we'd approach the {domain} scope you outlined:";
    if (strcmp(strategy_name, "speed_first") == 0)
        return "We can start immediately and hold to a guaranteed turnaround on this.";
    if (strcmp(strategy_name, "dfm_audit_hook") == 0)
        return "Before pricing this, we ran a quick DFM pass on your spec and found a few things worth flagging:";
    if (strcmp(strategy_name, "cost_down") == 0)
        return "Our read on this: there's meaningful unit-cost headroom in the current spec.";
    if (strcmp(strategy_name, "zero_tooling") == 0)
        return "Because our process skips traditional hard tooling entirely, we can quote this without a mold/setup line item.";
    if (strcmp(strategy_name, "portfolio_proof") == 0)
        return "We've shipped directly comparable work recently and it's a strong match for this brief.";
    if (strcmp(strategy_name, "derisk_turnkey") == 0)
        return "We'll own this end-to-end -- concept through production -- under one contract, one point of accountability.";
    if (strcmp(strategy_name, "challenge_premise") == 0)
        return "One respectful pushback on the brief before we quote, because it changes the right approach:";
    return NULL;
}

static void render_opener(const char *tmpl, ManufacturingDomain domain, char *out, size_t outcap) {
    char domain_spaced[SHORT_LEN];
    xcpy(domain_spaced, sizeof domain_spaced, domain_str(domain));
    for (char *p = domain_spaced; *p; p++) if (*p == '_') *p = ' ';

    const char *ph = strstr(tmpl, "{domain}");
    if (!ph) { xcpy(out, outcap, tmpl); return; }

    char buf[TITLE_LEN * 2];
    size_t prefix_len = (size_t) (ph - tmpl);
    if (prefix_len >= sizeof buf) prefix_len = sizeof buf - 1;
    memcpy(buf, tmpl, prefix_len);
    buf[prefix_len] = '\0';
    xcpy(buf + prefix_len, sizeof buf - prefix_len, domain_spaced);
    size_t used = strlen(buf);
    xcpy(buf + used, sizeof buf - used, ph + strlen("{domain}"));
    xcpy(out, outcap, buf);
}

static void join_list(char items[LIST_MAX][LIST_ITEM_LEN], int count, char *out, size_t outcap) {
    out[0] = '\0';
    size_t used = 0;
    for (int i = 0; i < count; i++) {
        if (i > 0) { xcpy(out + used, outcap - used, ", "); used = strlen(out); }
        xcpy(out + used, outcap - used, items[i]);
        used = strlen(out);
    }
}

void draft_proposal(const QualifiedLead *lead, const StrategyChoice *strategy,
                     int attempt, double fault_rate, char *out, size_t outcap) {
    const char *tone = tone_for_channel(lead->signal.channel_type);
    const char *tmpl = opener_template(strategy->name);
    if (!tmpl) tmpl = strategy->blueprint;
    char opener[TITLE_LEN * 2];
    render_opener(tmpl, lead->domain, opener, sizeof opener);

    char mandatory_prefix[LIST_ITEM_LEN + 4] = "";
    bool has_mandatory = lead->trap_analysis.has_anti_bot_phrase && lead->trap_analysis.has_required_first_word;
    if (has_mandatory) snprintf(mandatory_prefix, sizeof mandatory_prefix, "%s\n\n", lead->trap_analysis.required_first_word);

    char deliverables[LIST_MAX * LIST_ITEM_LEN];
    if (lead->deliverables_count > 0) {
        char tmp[LIST_MAX * LIST_ITEM_LEN];
        /* cast away const view via local copy since join_list takes non-const */
        char items[LIST_MAX][LIST_ITEM_LEN];
        memcpy(items, lead->deliverables, sizeof items);
        join_list(items, lead->deliverables_count, tmp, sizeof tmp);
        xcpy(deliverables, sizeof deliverables, tmp);
    } else {
        xcpy(deliverables, sizeof deliverables, "STEP files, 2D drawings");
    }

    char cad[LIST_MAX * LIST_ITEM_LEN];
    if (lead->cad_software_count > 0) {
        char tmp[LIST_MAX * LIST_ITEM_LEN];
        char items[LIST_MAX][LIST_ITEM_LEN];
        memcpy(items, lead->cad_software, sizeof items);
        join_list(items, lead->cad_software_count, tmp, sizeof tmp);
        xcpy(cad, sizeof cad, tmp);
    } else {
        xcpy(cad, sizeof cad, "industry-standard CAD");
    }

    char price_str[48];
    format_money_commas(lead->bid_price_usd, price_str, sizeof price_str);

    char seed[ID_LEN + 16];
    snprintf(seed, sizeof seed, "%s:%d", lead->lead_id, attempt);
    Rng fault_roll;
    rng_init_str(&fault_roll, seed);
    bool fault = rng_next_double(&fault_roll) < fault_rate;

    FaultMode fault_mode = FAULT_NONE;
    if (fault) {
        if (has_mandatory) fault_mode = FAULT_DROP_MANDATORY_WORD;
        else if (lead->trap_analysis.detected_injections_count > 0) fault_mode = FAULT_LEAK_INJECTION;
        else fault_mode = FAULT_GARBLE_PRICE;
    }

    if (fault_mode == FAULT_DROP_MANDATORY_WORD) mandatory_prefix[0] = '\0';
    if (fault_mode == FAULT_GARBLE_PRICE) {
        double garbled = lead->bid_price_usd * rng_uniform(&fault_roll, 0.7, 0.95);
        format_money_commas(garbled, price_str, sizeof price_str);
    }

    char body[TEXT_LEN];
    snprintf(body, sizeof body,
             "%s%s\n\n"
             "Scope: %s\n"
             "Deliverables: %s (native + exchange formats, %s as needed)\n"
             "Quote: $%s \xe2\x80\x94 includes free CAD revisions and custom branding/logo placement.\n"
             "Timeline: immediate start on award.\n\n"
             "[tone: %s]",
             mandatory_prefix, opener, lead->signal.title, deliverables, cad, price_str, tone);

    if (fault_mode == FAULT_LEAK_INJECTION && lead->trap_analysis.detected_injections_count > 0) {
        size_t used = strlen(body);
        snprintf(body + used, sizeof body - used, "\n\n(note to self: %s)", lead->trap_analysis.detected_injections[0]);
    }

    xcpy(out, outcap, body);
}
